### PR TITLE
luci-proto-ipip: Add package

### PR DESCRIPTION
**Pull request by rogerpueyo**
This package adds LuCI support for IPIP tunnels (IPv4-in-IPv4 RFC2003)

Signed-off-by: Roger Pueyo Centelles roger.pueyo@guifi.net

### CODE REVIEW

File: protocols/luci-proto-ipip/Makefile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# Copyright (C) 2008-2016 The LuCI Team <luci@lists.subsignal.org>
+#
+# This is free software, licensed under the Apache License, Version 2.0 .
+#
+
+include $(TOPDIR)/rules.mk
+
+LUCI_TITLE:=Support for IPIP tunnels (IPv4-in-IPv4 RFC2003)
+LUCI_DEPENDS:=
+
```
Comment (chris5560) [inline protocols/luci-proto-ipip/Makefile]: you should add:
LUCI_PKGARCH:=all

PKG_NAME:=.....
PKG_VERSION:=....
PKG_RELEASE=.....
PKG_MAINTAINER:=.....
PKG_LICENSE:=Apache-2.0

File: protocols/luci-proto-ipip/luasrc/model/network/proto_ipip.lua [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	return "ipip-" .. self.sid
+end
+
+function proto.opkg_package(self)
+	return "ipip"
+end
+
+function proto.is_installed(self)
+	return nixio.fs.access("/lib/netifd/proto/ipip.sh")
+end
```
Comment (chris5560) [inline protocols/luci-proto-ipip/luasrc/model/network/proto_ipip.lua]: inside Makefile there is no dependency definition.
Which package install ipip.sh ?
If someone selects your LuCI package for installation it should also install the needed package for ipip protocol.
ipip protocol is not a standard protocol of OpenWrt build

**Comment by chris5560** [comment]
@rogerpueyo 
Sorry for the delay.

**Comment by rogerpueyo** [comment]
@chris5560 
Thanks for your comments! The current commit should do the trick regarding the Makefile. No worries about the delay :)

File: protocols/luci-proto-ipip/Makefile [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#
+# Copyright (C) 2008-2016 The LuCI Team <luci@lists.subsignal.org>
```
Comment (jow-) [inline protocols/luci-proto-ipip/Makefile]: Please use your own copyright here or don't set any.

File: protocols/luci-proto-ipip/Makefile [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+# This is free software, licensed under the Apache License, Version 2.0 .
+#
+
+include $(TOPDIR)/rules.mk
+
+LUCI_TITLE:=Support for IPIP tunnels (IPv4-in-IPv4 RFC2003)
+LUCI_DEPENDS:=+ipip
+LUCI_PKGARCH:=all
+
+PKG_NAME:=luci-proto-ipip
```
Comment (jow-) [inline protocols/luci-proto-ipip/Makefile]: This is set by default from `luci.mk`

File: protocols/luci-proto-ipip/Makefile [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+PKG_RELEASE=1
```
Comment (jow-) [inline protocols/luci-proto-ipip/Makefile]: This is not needed for LuCI modules

File: protocols/luci-proto-ipip/Makefile [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+PKG_VERSION:=1.0.0
```
Comment (jow-) [inline protocols/luci-proto-ipip/Makefile]: This is automatically inferred from the Git repo version by `luci.mk`

File: protocols/luci-proto-ipip/Makefile [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (C) 2008-2016 The LuCI Team <luci@lists.subsignal.org>
+#
+# This is free software, licensed under the Apache License, Version 2.0 .
+#
+
+include $(TOPDIR)/rules.mk
+
+LUCI_TITLE:=Support for IPIP tunnels (IPv4-in-IPv4 RFC2003)
+LUCI_DEPENDS:=+ipip
+LUCI_PKGARCH:=all
```
Comment (jow-) [inline protocols/luci-proto-ipip/Makefile]: Package arch `all` is the default if there is no `src/Makefile` present in the module

File: protocols/luci-proto-ipip/Makefile [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+PKG_NAME:=luci-proto-ipip
+PKG_RELEASE=1
+PKG_VERSION:=1.0.0
+PKG_MAINTAINER:=Roger Pueyo Centelles <roger.pueyo@guifi.net>
+PKG_LICENSE:=Apache-2.0
```
Comment (jow-) [inline protocols/luci-proto-ipip/Makefile]: All of LuCI is Apache 2 so this is superflous

File: protocols/luci-proto-ipip/Makefile [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+include $(TOPDIR)/rules.mk
+
+LUCI_TITLE:=Support for IPIP tunnels (IPv4-in-IPv4 RFC2003)
+LUCI_DEPENDS:=+ipip
+LUCI_PKGARCH:=all
+
+PKG_NAME:=luci-proto-ipip
+PKG_RELEASE=1
+PKG_VERSION:=1.0.0
+PKG_MAINTAINER:=Roger Pueyo Centelles <roger.pueyo@guifi.net>
```
Comment (jow-) [inline protocols/luci-proto-ipip/Makefile]: This can stay :)

File: protocols/luci-proto-ipip/luasrc/model/cbi/admin_network/proto_ipip.lua [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+ipaddr.datatype = "ip4addr"
+
+tunlink = section:taboption("general", Value, "tunlink", translate("Bind interface"), translate("Bind the tunnel to this interface (optional)."))
+ipaddr.optional = true
+
+
+mtu = section:taboption("advanced", Value, "mtu", translate("Override MTU"), translate("Specify an MTU (Maximum Transmission Unit) other than the default (1280 bytes)."))
+mtu.optional = true
+mtu.placeholder = 1280
+mtu.datatype = "and(min(68),max(9200))"
```
Comment (jow-) [inline protocols/luci-proto-ipip/luasrc/model/cbi/admin_network/proto_ipip.lua]: You can use `range(68, 9200)` here

File: protocols/luci-proto-ipip/luasrc/model/cbi/admin_network/proto_ipip.lua [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+ttl = section:taboption("advanced", Value, "ttl", translate("Override TTL"), translate("Specify a TTL (Time to Live) for the encapsulating packet other than the default (64)."))
+ttl.optional = true
+ttl.placeholder = 64
+ttl.datatype = "min(1)"
+
+tos = section:taboption("advanced", Value, "tos", translate("Override TOS"), translate("Specify a TOS (Type of Service)."))
+tos.optional = true
+tos.datatype = "and(min(0),max(255))"
```
Comment (jow-) [inline protocols/luci-proto-ipip/luasrc/model/cbi/admin_network/proto_ipip.lua]: Use `range(0, 255)` here

File: protocols/luci-proto-ipip/Makefile [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+PKG_LICENSE:=Apache-2.0
+
+LUA_TARGET:=source
```
Comment (jow-) [inline protocols/luci-proto-ipip/Makefile]: This variable is not used by `luci.mk` - please remove it.

**Comment by rogerpueyo** [comment]
@jow-,
Thanks for the review. The requested changes have been made. :)

File: protocols/luci-proto-ipip/Makefile (hunk 3 above)
Comment (rogerpueyo) [inline protocols/luci-proto-ipip/Makefile]: Changed to my own copyright.

File: protocols/luci-proto-ipip/Makefile (hunk 5 above)
Comment (rogerpueyo) [inline protocols/luci-proto-ipip/Makefile]: Removed.

File: protocols/luci-proto-ipip/Makefile (hunk 6 above)
Comment (rogerpueyo) [inline protocols/luci-proto-ipip/Makefile]: Removed.

File: protocols/luci-proto-ipip/Makefile (hunk 9 above)
Comment (rogerpueyo) [inline protocols/luci-proto-ipip/Makefile]: ACK.

File: protocols/luci-proto-ipip/Makefile (hunk 8 above)
Comment (rogerpueyo) [inline protocols/luci-proto-ipip/Makefile]: Removed.

File: protocols/luci-proto-ipip/Makefile (hunk 12 above)
Comment (rogerpueyo) [inline protocols/luci-proto-ipip/Makefile]: Removed.

File: protocols/luci-proto-ipip/luasrc/model/cbi/admin_network/proto_ipip.lua (hunk 10 above)
Comment (rogerpueyo) [inline protocols/luci-proto-ipip/luasrc/model/cbi/admin_network/proto_ipip.lua]: Changed.

File: protocols/luci-proto-ipip/luasrc/model/cbi/admin_network/proto_ipip.lua (hunk 11 above)
Comment (rogerpueyo) [inline protocols/luci-proto-ipip/luasrc/model/cbi/admin_network/proto_ipip.lua]: Changed.

**Comment by rogerpueyo** [comment]
@jow-,
Thanks for the review. The requested changes have been made. :) Please consider the commit for upstream merging.

**Comment by danrl** [comment]
I would like to know why this PR is not yet merged, becausd I have a PR waiting myself and would like to speed things up by learning from other PRs.

**Comment by rogerpueyo** [comment]
@danrl I made the requested changes but I guess they were unnoticed, or nobody has had time to merge it yet :disappointed:

**Comment by danrl** [comment]
I see 😥

**Comment by rogerpueyo** [comment]
Dear @jow- @chris5560,
I wonder if the commit is ready for upstream merging. After fixing the points you raised, I've also updated the code to this related commit: https://github.com/lede-project/source/commit/311682905edb56abd31c7ece3664910586070b91
Thanks!

**Comment by chris5560** [comment]
@rogerpueyo
Sorry for the delay.
Does LEDE project merge the dependent package ?

**Comment by rogerpueyo** [comment]
@chris5560 
No problem. I didn't see your comment from the 4th October, my apologies.

> inside Makefile there is no dependency definition.
> Which package install ipip.sh ?
> If someone selects your LuCI package for installation it should also install the needed package for ipip protocol.
> ipip protocol is not a standard protocol of OpenWrt build

I understand that the line `LUCI_DEPENDS:=+ipip` adds the "ipip" package as a dependency, doesn't it?
The ipip package is shipped by LEDE (also by OpenWrt) and it is the one that includes the ipip.sh file (actually, it contains only that file: https://github.com/lede-project/source/tree/master/package/network/config/ipip).

**Pull request merged**